server/src/domain/storage/storage.core.ts

```diff
@@ -42,6 +51,33 @@ export class StorageCore {
     return join(folderPath, fileName);
   }
 
+  async moveFile(id: string, pathType: PathType, originalPath: string, newPath: string) {
+    const move = await this.moveRepository.create(id, pathType, originalPath, newPath);
+    await this.repository.moveFile(originalPath, newPath);
```

Comment:
Some consideration for race conditions needs to be had here. What happens when a request comes in during this process? So, the file has been moved, but the asset hasn't been updated? It will fail, right? Either the row needs to be locked for the duration of this operation, or the code which loads assets needs to handle this. Even additional logic for checking if the file exists can be racy, as the file may have then been moved after the check. I guess this would be fine if the file handle is used directly as move operations don't affect inodes.

Reply:
One option we had considered before is linking the inode to the new path and the old path at the same time, but this is not supported by some filesystems like rclone fuse.

Reply:
Though, I am pretty sure users would not need to use fuse filesystems if object storage is supported natively.

Reply:
I'm unclear on what your concern here is - errors like the asset table not getting the new path are exactly the scenario that this new move table is meant to address. Can you clarify?

Reply:
There can still be race conditions writing old/new paths to the table, so it moves the race conditions from the move file operation to the database insert. If we add a unique constraint that prevents two pending move records from being created for the same file then we'd eliminate that race condition I believe.

Reply:
Well, not even just that. Break this entire operation down step-by-step and imagine each database query takes an excessively long time (1 second).

At current, the code will move the file and then update the paths on the asset row. So, imagine the file has been moved, but the asset has not yet been updated. A request to load the asset will now fail, because the asset loading code will try to use the old path, which doesn't exist.

Reply:
The failure scenario you describe is valid, but should be temporary and recoverable. Not ideal, but at least much better than what we have now.

Reply:
I don't think this is something which should be glossed over - there is no point building a system with known failure cases like this.

Reply:
I think it is hard to address that issue now, but we can do it more easily in the future. The current system has this limitation too. Right now the "read and stream back file" logic is scattered. Eventually we can have something like `storageCore.sendFile(path)`, which can specifically handle file not found errors and check the pending move table in order to get the old path and test it.

Reply:
Then what is the point of this PR if it doesn't make the storage robust? There are still clear issues and the system can end up in exactly the same state as without the changes. Rerunning the storage migration job already sort of resolves the original issue, so if that is still required then this basically changes nothing?

Reply:
In the current system, rerunning the job cannot always resolve the issue as it is possible for a file to become 'lost'. That is what this PR seems to address.

Reply:
It just seems like a wasted effort to introduce all of this complexity and not even fully resolve the issue.